File: javascript_executor/src/services/executor.service.js
'use strict'

const fs = require('fs')
const path = require('path')
const { exec } = require('child_process');

class ExecutorService {
    // 1. Create folder to store code files
    // 2. Hash name of file based on user, problem and create file
    // 3. Write code to file and execute
    async execute({ code, userId, problemId }) {
        // 1
        const dirPath = path.join(__dirname, '../temp')
        if (!fs.existsSync(dirPath)) {
            fs.mkdirSync(dirPath)
        }

        // 2
        const filePath = path.join(dirPath, `${userId}-${problemId}.js`)
        
        // 3
        return new Promise((resolve, reject) => {
            fs.writeFile(filePath, code, (err) => {
                if (err) {
                    console.error('Error writing file:', err.message)
                    return reject(err.message)
                }

                exec(`node ${filePath}`, (err, stdout, stderr) => {
                    if (err) {
                        return reject(stderr)
                    }

                    return resolve(stdout || stderr)
                })
            })
        }).then((data) => {
            return data.trim()
        }).catch((err) => {
            throw err
        }).finally(() => {
            // Delete file after execution
            fs.unlinkSync(filePath)
        })
    }
}

module.exports = new ExecutorService()

File: java_executor/src/main/java/com/executor/java_executor/services/ExecutorService.java
package com.executor.java_executor.services;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.CompletableFuture;

import org.springframework.stereotype.Service;

import com.executor.java_executor.dtos.requests.ExecuteCodeRequest;

@Service
public class ExecutorService {
    // 1. Create temp directory
    // 2. Create file path with userId and problemId
    // 3. Write code to file
    // 4. Run code using ProcessBuilder
    public CompletableFuture<String> run(ExecuteCodeRequest executeCodeRequest) throws Exception {
        String code = executeCodeRequest.getCode();
        String userId = executeCodeRequest.getUserId();
        String problemId = executeCodeRequest.getProblemId();
        // 1
        String dirPath = "temp" + File.separator + userId + "_" + problemId;
        Path tempDirPath = Paths.get(dirPath);

        try {
            Files.createDirectories(tempDirPath);
        } catch (Exception e) {
            throw new RuntimeException("Error creating temp directory: " + e.getMessage());
        }

        // 2
        String fileName = "Main.java";
        Path filePath = tempDirPath.resolve(fileName);
        File codeFile = new File(filePath.toString());

        // 3
        try (FileWriter writer = new FileWriter(codeFile)) {
            writer.write(code);
        } catch (Exception e) {
            throw new RuntimeException("Error writing code to file: " + e.getMessage());
        }

        // 4
        ProcessBuilder processBuilder = new ProcessBuilder("java", filePath.toString());
        processBuilder.directory(new File(System.getProperty("user.dir")));

        return CompletableFuture.supplyAsync(() -> {
            StringBuilder result = new StringBuilder();
            try {
                Process process = processBuilder.start();
                int exitCode = process.waitFor();
                
                // Get stdout and stderr
                result.append(new String(process.getInputStream().readAllBytes()));
                result.append(new String(process.getErrorStream().readAllBytes()));

                // Return result or error output
                if (exitCode != 0) {
                    throw new RuntimeException("Error executing code: " + result.toString());
                }
            } catch (Exception e) {
                throw new RuntimeException("Execution failed: " + e.getMessage());
            } finally {
                // Delete the file after execution
                try {
                    Files.deleteIfExists(filePath);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            return result.toString().trim();
        });
    }
}
